Python 3 compatibility for SWC website build tools

wking opened the pull request:
With these patches,

  $ make check

succeeds using Python 3.  Python 2.{6,7} should also work, but earlier
versions (e.g. Python 2.5) will choke on some newer syntax.

wking committed:
compile.py: use keys instead of cmp for sorting

cmp is gone in Python 3.

I use the tuple (x._sort_key, id(x)) instead of just x._sort_key to
avoid

  Traceback (most recent call last):
    ...
    File "bin/compile.py", line 248, in _load_subfiles
      self.children.sort(key=lambda x: x._sort_key)
  TypeError: unorderable types: NoneType() < NoneType()

which happens when multiple children have _sort_key == None.  Breaking
these ties based on the child's id is random, but it works.  If you
want a predictable tie breaker, fix _sort_key ;).

wking committed:
compile.py: replace "print >> sys.stderr, ..." with "stderr.write(...)"

"print >>" is gone in Python 3.

wking committed:
validxml.py: fix "except a, b" to "except a as b"

Updates to Python 3 compatibility (PEP 3110).  The new syntax was
backported to Python 2.6 and 2.7.

wking committed:
validxml.py: convert "'%s' % (...)" -> "'{0}'.format(...)"

The '' % () syntax is gone in Python 3.  Python 2.6+ supports
.format().

wking committed:
validxml.py: adjust entitydefs import

The htmlentitydefs module has been renamed to html.entities in Python 3.

wking committed:
links.py: convert "'%s' % (...)" -> "'{0}'.format(...)"

The '' % () syntax is gone in Python 3.  Python 2.6+ supports
.format().

wking committed:
util.py: fix "except a, b" to "except a as b"

Updates to Python 3 compatibility (PEP 3110).  The new syntax was
backported to Python 2.6 and 2.7.

wking committed:
util.py: convert "'%s' % (...)" -> "'{0}'.format(...)"

The '' % () syntax is gone in Python 3.  Python 2.6+ supports
.format().

wking committed:
util.py: adjust entitydefs import

The htmlentitydefs module has been renamed to html.entities in Python 3.

wking committed:
chars.py: convert "'%s' % (...)" -> "'{0}'.format(...)"

The '' % () syntax is gone in Python 3.  Python 2.6+ supports
.format().

wking committed:
journal.py: replace "print ..." with "print(...)"

The "print ..." statement was removed in Python 3.  Use the print()
function instead, which has been around since at least Python 2.4.
See PEP 3105 for the motivation for removing the old statement.

wking committed:
util.py: adjust StringIO import for Python 3 compatibility

Assorted *IO classes have been consolidated into the "io" package in
Python 3.

gvwilson commented:
Do you have access to a Python 2.6.6 to test?  That's what we've got on the server right now.

wking committed:
links.py: Don't use set literals for Python 2.6.5 compatibility.

I'm not sure exactly when set literals were backported to the Python 2
series, but:

  $ python --version
  Python 2.6.5
  $ python -c '{1, 2}'
    File "<string>", line 1
      {1, 2}
        ^
  SyntaxError: invalid syntax

wking committed:
util.py: Work around ElementTree 1.2 issues

If you're running a very old version of Python (like 2.6.5), you may
not have:

* xml.etree.ElementTree.ParseError
* xml.etree.ElementTree.XMLParser().parser

Work around that (on the old versions only) by using ExpatError as a
substiture for ParseError.  Also skip foreign DTD configuration if
parser.parser is missing.

wking committed:
links.py: Work around XPath attributes for ElementTree 1.2

ElementTree 1.2 doesn't support XPath attributes, so we need to filter
out tags with hrefs ourselves.

wking commented:
On Mon, Dec 17, 2012 at 06:45:39AM -0800, Greg Wilson wrote:

> Do you have access to a Python 2.6.6 to test?  That's what we've got
> on the server right now.

I don't have access to 2.6.6, but I just tested it on 2.6.5.  It works
with 2.6.5, but I had to add three more patches to get 2.6.5
compatibility.  The patches (5f6154ac, a92101ec, and 1a82a291) are
syntax and ElementTree fixes for the old Python, and aren't related to
my Python 3 compatibility fixes.  Feel free to drop the new patches if
you're not interested in keeping compatibility with 2.6.5.

gvwilson commented:
Thanks very much for this --- ElementTree changed in 2.7, so if it works 
with 2.6.5, it'll work with 2.6.6. In theory.  I've put in a change req 
to Dreamhost for a Python upgrade, but I'm not holding my breath.

On 2012-12-17 11:52 AM, W. Trevor King wrote:

> On Mon, Dec 17, 2012 at 06:45:39AM -0800, Greg Wilson wrote:
> 
> > Do you have access to a Python 2.6.6 to test? That's what we've got
> > on the server right now.
> 
> I don't have access to 2.6.6, but I just tested it on 2.6.5. It works
> with 2.6.5, but I had to add three more patches to get 2.6.5
> compatibility. The patches (5f6154ac, a92101ec, and 1a82a291) are
> syntax and ElementTree fixes for the old Python, and aren't related to
> my Python 3 compatibility fixes. Feel free to drop the new patches if
> you're not interested in keeping compatibility with 2.6.5.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/swcarpentry/website/pull/118#issuecomment-11449517.
